components/omega/src/ocn/TracerDefs.inc

```diff
@@ -0,0 +1,25 @@
+//===-- ocn/TracerDefs.inc --------------------*- C++ -*-===//
+//
+/// \file
+/// \brief OMEGA Tracer Definition File
+///
+/// This file defines all OMEGA tracers. It is imported using
+/// the C++ #include preprocessor directive in Tracers.cpp. Only the tracer
+/// definitions selected in the Tracers section of the OMEGA YAML configuration
+/// file will be allocated into memory.
+//===----------------------------------------------------------------------===//
+
```

Comment:
One of the reasons for the getByIndex function was to avoid the more expensive string comparison in getByName. In the current implementation, we would still need to get the index by name first and then use the index. I'm wondering whether in the include file, we could include a list of shared, public indices for each tracer (set when they are defined). Like IndxTemp, IndxSalt, IndxDebug1, etc.  Since we only want to edit one file when adding tracers, that would mean putting the indices at the top of this include file and including it in Tracers.h. That means we would have to wrap all the defines in a defineAll function that is called where the current inc file is included.

What do others think?

Reply:
@philipwjones , what are the benefits of having a list of shared and public indices for each tracers?

Reply:
By way of explanation, this is what I'm thinking.  The TracersDef.inc file would look like:

```c++
// Index defined for each tracer. The value of these tracer indices are still set in Tracers::init()  and
// tracers that have not been selected would have standard invalid value: Tracer::IndxInvalid = -1
static int IndxTemp;
static int IndxSalt;
static int IndxMyBGCTracer;
static int IndxAnotherBGCTracer;
static int IndxDebug1;
static int IndxDebug2;
etc. for all defined tracers

// Tracer definitions packaged in a defineAllTracers function
static void defineAllTracers() {

  // here we put all the define calls as they exist in the current TracersDef.inc...
}
```
The TracersDef.inc file would then be included in the Tracers.h file in the public part of the Tracer class definition so that they are all available using eg Tracer::IndxSalt.  In the Tracers.cpp file where the include currently sits, we would replace the include statement with a call to defineAllTracers.  The init function would still assign consecutive indices to groups.

This would allow everyone to access the indices directly, providing speedier access without an explicit call to search for the index by name. I realize this is more of a Fortran style and not the usual C++ approach but I think it is both simpler and faster than a search by name or having all modules do the lookup on init.

This is just a proposal - others should weigh in...

Reply:
@philipwjones, I tried to implement your suggested code in this branch (https://github.com/grnydawn/Omega/tree/ykim/omega/tracerinfra-def) but encountered a linking issue:

```
ld.lld: error: undefined symbol: OMEGA::Tracers::IndxTemp
>>> referenced by Tracers.cpp
>>>               Tracers.cpp.o:(OMEGA::Tracers::defineAllTracers()) in archive ../src/libOmegaLib.a
```

To resolve the linker error, I added the following near the top of `Tracers.cpp`:

```cpp
I4 Tracers::IndxTemp = 0;
I4 Tracers::IndxSalt = 0;
```

This fix worked, but it makes me think that an additional user-provided include file might be necessary. I’m still learning C++, so I may be missing something.

Reply:
That's right - forgot about that. The header just has the forward declaration but we still need to initialize in the implementation (.cpp) file.  We might be able to include the same file in both with some creative ifdefs but this is starting to get more complicated than I had intended. Let me think about it some more

Reply:
Let's stick with your initial implementation for now

Reply:
I am bit late to this conversation, but if you declare the member index variables to be `static inline` then you can initialize them in the header file. Like this:
```
inline static int IndxTemp = -1;
```

Reply:
Just to clarify, and echo a comment from @xylar at the last meeting: this is just a convenience/optimization that can be done for certain tracers and not a requirement, correct?

Reply:
@mwarusz, thanks for the tip about using inline. I have updated Tracers with the inline, and it worked well.

@philipwjones, I am using the variable names for the static indexes from your example code above. If you need to add more tracer index variable names in "TracerDefs.inc", let me know, or we can update the file later.